test/BulletDynamics/test_deterministic_save_and_restore.cpp

```diff
@@ -0,0 +1,221 @@
+
+	world->setGravity(btVector3(0, -10, 0));
+
+	return world;
+
+}
+
+GTEST_TEST(BulletDynamics, DeterministicSaveRestore)
+{
+
+	static char filename[] = "test_serialize.mixail";
+	btRigidBody* box_before[3];
+	btCollisionObject* box_after[3];
+	btMultiBodyDynamicsWorld *initial_world = create_btMultiBodyDynamicsWorld();
+
+	// create ground plane and box shapes
+	// don't worry about gui things
+
+	btBoxShape* groundShape = new btBoxShape(btVector3(btScalar(50.), btScalar(50.), btScalar(50.)));
+	{
+		btTransform groundTransform;
+		groundTransform.setIdentity();
+		groundTransform.setOrigin(btVector3(0, -50, 0));
+		btScalar mass(0.);
+		bool isDynamic = (mass != 0.f);
+		btVector3 localInertia(0, 0, 0);
+		if (isDynamic)
+			groundShape->calculateLocalInertia(mass, localInertia);
+
+		#ifdef USE_MOTIONSTATE
+			btDefaultMotionState* myMotionState = new btDefaultMotionState(groundTransform);
+
+			btRigidBody::btRigidBodyConstructionInfo cInfo(mass, myMotionState, groundShape, localInertia);
+
+			btRigidBody* body = new btRigidBody(cInfo);
+		#else
+			btRigidBody* body = new btRigidBody(mass, 0, groundShape, localInertia);
+			body->setWorldTransform(groundTransform);
+		#endif
+
+		body->setUserIndex(-1);
+		initial_world->addRigidBody(body);
+	}
+	{
+
+		btBoxShape* colShape = new btBoxShape(btVector3(.1, .1, .1));
+
+		btTransform startTransform;
+		startTransform.setIdentity();
+
+		btScalar mass(1.f);
+
+		bool isDynamic = (mass != 0.f);
+
+		btVector3 localInertia(0, 0, 0);
+		if (isDynamic)
+			colShape->calculateLocalInertia(mass, localInertia);
+		for (int i = 0; i < 3; i++)
+		{
+			int a[4];
+			if (i == 0)
+			{
+				a[0] = 2;
+				a[1] = 0;
+				a[2] = -1;
+				a[3] = 1;
+			}
+			else if (i ==1)
+			{
+				a[0] = 2;
+				a[1] = 2;
+				a[2] = -1;
+				a[3] = -1;
+			}
+			else
+			{
+				a[0] = 0;
+				a[1] = 1;
+				a[2] = 1;
+				a[3] = 0;
+			}
+			startTransform.setOrigin(btVector3(
+				btScalar(a[0]),
+				btScalar(10),
+				btScalar(a[1])));
+			#ifdef USE_MOTIONSTATE
+				btDefaultMotionState* myMotionState = new btDefaultMotionState(startTransform);
+
+				btRigidBody::btRigidBodyConstructionInfo cInfo(mass, myMotionState, colShape, localInertia);
+
+				box_before[i] = new btRigidBody(cInfo);
+			#else
+				box_before[i] = new btRigidBody(mass, 0, colShape, localInertia);
+				box_before[i]->setWorldTransform(startTransform);
+			#endif
+			box_before[i]->setUserIndex(-1);
+			initial_world->addRigidBody(box_before[i]);
+			box_before[i]->setLinearVelocity(btVector3(a[2], 0, a[3]));
+		}
+	}
+
+	// step simulation until collision
+
+	const double delta_t = 0.01;
+	int time = 0;
+	while (box_before[0]->getCenterOfMassPosition().getY() == box_before[1]->getCenterOfMassPosition().getY() &&
+			box_before[1]->getCenterOfMassPosition().getY() == box_before[2]->getCenterOfMassPosition().getY())
+	{
+		time++;
+		initial_world->stepSimulation(delta_t);
+	}
```

Comment:
I see, you're using the fact that the blocks start in the same y position and all start with 0 y-velocity to determine collision. Since what we're actually interested in is whether or not there are any collision manifolds active, let's check for that. You can do this by checking with the dispatcher, i.e. initial_world->getDispatcher()->getNumManifolds(). We should make sure that there are at least 2 collision manifolds present for multiple time steps, because part of the non-determinism comes from collision manifolds that add points to the contact manifold over multiple time steps.